File: SixFour/Palette/IsometryMove.swift
import Foundation
import simd

/// The EXACT delta-preserving move on Q16 OKLab — the Swift port of `SixFour.Spec.IsometryMove`.
///
/// A lattice isometry = per-axis sign-flips (each ±1 on L,a,b) + an integer Q16 translation. These
/// are the ONLY moves that preserve every pairwise OKLab distance with NO tolerance (the translation
/// cancels in any difference; sign² = 1). So applying ONE move to a whole palette shifts every colour
/// coherently — the relative deltas WITHIN a frame cannot change at all, and applied identically to
/// all 64 frames, neither can the deltas BETWEEN frames. Pure integer negate+add per channel ⇒
/// SIMD-native. Verified bit-for-bit against the Haskell laws in `IsometryMoveTests`.
struct IsoMove: Equatable {
    /// Per-axis ±1 sign-flip on (L, a, b).
    var signs: SIMD3<Int32>
    /// Q16 translation added after the flip.
    var shift: SIMD3<Int32>

    static let identity = IsoMove(signs: SIMD3(1, 1, 1), shift: .zero)
    /// σ = (L, −a, −b) — the canonical reflection (`PairTree.lawSigmaEuclideanIsometry`).
    static let sigma = IsoMove(signs: SIMD3(1, -1, -1), shift: .zero)
    /// A pure translation — the continuous A/B knob.
    static func translate(_ t: SIMD3<Int32>) -> IsoMove { IsoMove(signs: SIMD3(1, 1, 1), shift: t) }

    /// Apply the move to one colour: flip each axis, then add the translation.
    func apply(_ c: SIMD3<Int32>) -> SIMD3<Int32> { signs &* c &+ shift }

    /// The inverse move (`apply(invert.apply(c)) == c`): re-use the signs, translate by −(s·t).
    var inverse: IsoMove { IsoMove(signs: signs, shift: SIMD3<Int32>.zero &- (signs &* shift)) }

    /// Squared Q16 distance (Int, to avoid Int32 overflow on the ~1.3e5 max difference).
    static func distSq(_ a: SIMD3<Int32>, _ b: SIMD3<Int32>) -> Int {
        let dx = Int(a.x) - Int(b.x), dy = Int(a.y) - Int(b.y), dz = Int(a.z) - Int(b.z)
        return dx * dx + dy * dy + dz * dz
    }
}

/// The Swift port of `SixFour.Spec.MoveRadiusSchedule` — anneal the move magnitude (wide early →
/// JND floor) + a hard cumulative-displacement cap, all exact Q16. The "visible reload without drift"
/// schedule: early rounds move a visible amount, late rounds a JND, the look never drifts past ±0.25
/// OKLab from the capture. Verified against the Haskell laws in `IsometryMoveTests`.
enum MoveRadiusSchedule {
    static let radiusMax: Int32 = 8192   // ≈ 0.125 OKLab — cold-start (A/B visibly different)
    static let radiusMin: Int32 = 1024   // ≈ 0.0156 (one JND) — every round still moves
    static let halfLife: Int32 = 8       // picks to halve the excess radius
    static let cumCap: Int32 = 16384     // ≈ 0.25 — hard per-axis cap on cumulative drift

    /// Per-round move radius (Q16): `rMin + (rMax−rMin)·halfLife / (n+halfLife)`. Integer division
    /// ⇒ exact + deterministic. `rMax` at n=0, monotone ↓ to `rMin`.
    static func radius(_ n: Int) -> Int32 {
        let nn = Int32(max(0, n))
        return radiusMin + ((radiusMax - radiusMin) * halfLife) / (nn + halfLife)
    }

    /// Project a cumulative displacement into the L∞ ball of radius `cumCap` — the chosen look can
    /// never drift more than `cumCap` from the original capture, however many rounds accrue.
    static func clampToCap(_ t: SIMD3<Int32>) -> SIMD3<Int32> {
        SIMD3(min(cumCap, max(-cumCap, t.x)),
              min(cumCap, max(-cumCap, t.y)),
              min(cumCap, max(-cumCap, t.z)))
    }
}

File: spec/src/SixFour/Spec/IsometryMove.hs
{- |
Module      : SixFour.Spec.IsometryMove
Description : The EXACT, SIMT-native delta-preserving move on Q16 OKLab — the A/B genome step.

The A/B genome game advances both candidates by a /structure-preserving/ transform, so the
relative colour deltas WITHIN a frame — and, applied identically to every frame, BETWEEN
frames — do not change (the cause of the degradation was an unconstrained per-leaf nudge that
accumulated). On the integer Q16 lattice the moves that preserve every pairwise squared
distance EXACTLY (no ε) are the lattice isometries built from:

  * an integer TRANSLATION @t@ — the continuous A\/B knob (a global OKLab shift); and
  * an axis SIGN-FLIP @s ∈ {+1,−1}³@ on (L,a,b), of which σ = (L,−a,−b)
    (the proven 'SixFour.Spec.PairTree.lawSigmaEuclideanIsometry') is the canonical member.

Each is a handful of integer negate\/add ops per colour — byte-exact AND embarrassingly
SIMD\/SIMT-parallel across the 256 leaves × 64 frames (no matmul, no rounding). A continuous
rotation is deliberately EXCLUDED: a rounded integer rotation is reversible but drifts ~1 LSB
per shear, so it cannot carry a no-tolerance pairwise-delta law.

== The move

@
'apply' ('IsoMove' (sₗ,sₐ,s_b) (tₗ,tₐ,t_b)) (l,a,b) = (sₗ·l+tₗ, sₐ·a+tₐ, s_b·b+t_b)
@

The composition of a sign-flip and a translation is a lattice isometry; 'applyMove' maps it
over a whole list of generators, moving every colour /coherently/ — only a global reflect +
shift, never an independent per-leaf nudge.

Laws (QuickCheck'd in @Properties.IsometryMove@, EXACT — no ε):

  * 'lawMovePreservesPairwiseDelta' — @distSqQ16 (apply m x) (apply m y) == distSqQ16 x y@.
  * 'lawMoveReversible' — @apply (invert m) (apply m x) == x@.
  * 'lawIdentityIsNoOp' — the identity move is a byte-exact no-op.
  * 'lawSigmaIsAMove' — @apply sigmaMove (l,a,b) == (l,−a,−b)@ (σ as an 'IsoMove').
  * 'lawComposePreservesPairwiseDelta' — the move group is closed (composition stays exact).

GHC-boot-only. The Swift\/Zig port is one integer negate+add per channel per lane (SIMT).
-}
-- COMPARTMENT: METAL-GPU | tag:CommitSide
module SixFour.Spec.IsometryMove
  ( -- * The move
    IsoMove(..)
  , Sign
  , apply
  , applyMove
  , invert
  , compose
    -- * Canonical instances
  , identityMove
  , translateMove
  , sigmaMove
    -- * Laws (QuickCheck'd in Properties.IsometryMove, EXACT — no ε)
  , lawMovePreservesPairwiseDelta
  , lawMoveReversible
  , lawIdentityIsNoOp
  , lawSigmaIsAMove
  , lawComposePreservesPairwiseDelta
  ) where

import SixFour.Spec.PairTreeFixed (OKLabI)
import SixFour.Spec.QuantFixed    (distSqQ16)

-- | An axis sign: @+1@ (keep) or @−1@ (flip). Constrained to @{+1,−1}@; the generators in
-- @Properties.IsometryMove@ only ever produce @±1@ (any other value breaks the isometry laws).
type Sign = Int

-- | A lattice isometry on Q16 OKLab: per-axis sign-flips @(sₗ,sₐ,s_b)@ (each @±1@) followed by
-- an integer translation @(tₗ,tₐ,t_b)@. Both halves preserve every pairwise distance exactly,
-- so their composition does too.
data IsoMove = IsoMove
  { isoSigns :: (Sign, Sign, Sign)   -- ^ per-axis @±1@ sign-flip on (L,a,b)
  , isoShift :: OKLabI               -- ^ integer Q16 translation added after the flip
  } deriving (Eq, Show)

-- | Apply the move to one colour: flip each axis by its sign, then add the translation.
apply :: IsoMove -> OKLabI -> OKLabI
apply (IsoMove (sl, sa, sb) (tl, ta, tb)) (l, a, b) =
  (sl * l + tl, sa * a + ta, sb * b + tb)

-- | Apply the move to every colour in a list (the σ-pair generators / a palette) — the SAME
-- isometry to all, so their relative structure is preserved.
applyMove :: IsoMove -> [OKLabI] -> [OKLabI]
applyMove = map . apply

-- | The inverse move (@apply (invert m) . apply m == id@). Each sign is its own inverse, so the
-- inverse re-uses the signs and translates by @−(s·t)@.
invert :: IsoMove -> IsoMove
invert (IsoMove ss@(sl, sa, sb) (tl, ta, tb)) =
  IsoMove ss (negate (sl * tl), negate (sa * ta), negate (sb * tb))

-- | Compose two moves into one (@apply (compose m2 m1) == apply m2 . apply m1@). The result is
-- again a sign-flip + translation, so it remains an exact isometry.
compose :: IsoMove -> IsoMove -> IsoMove
compose (IsoMove (s2l, s2a, s2b) (t2l, t2a, t2b)) (IsoMove (s1l, s1a, s1b) (t1l, t1a, t1b)) =
  IsoMove (s2l * s1l, s2a * s1a, s2b * s1b)
          (s2l * t1l + t2l, s2a * t1a + t2a, s2b * t1b + t2b)

-- | The identity move (no flip, no shift) — a byte-exact no-op.
identityMove :: IsoMove
identityMove = IsoMove (1, 1, 1) (0, 0, 0)

-- | A pure translation (no flip) by a Q16 vector — the continuous A/B knob.
translateMove :: OKLabI -> IsoMove
translateMove = IsoMove (1, 1, 1)

-- | σ as a move: @(L, −a, −b)@ — the canonical reflection
-- ('SixFour.Spec.PairTree.lawSigmaEuclideanIsometry'), with no translation.
sigmaMove :: IsoMove
sigmaMove = IsoMove (1, -1, -1) (0, 0, 0)

-- | T1 — the move preserves EVERY pairwise squared distance exactly (the relative deltas within
-- a frame cannot change; applied identically to all frames, neither can the deltas between
-- frames). No ε.
lawMovePreservesPairwiseDelta :: IsoMove -> OKLabI -> OKLabI -> Bool
lawMovePreservesPairwiseDelta m x y =
  distSqQ16 (apply m x) (apply m y) == distSqQ16 x y

-- | T2 — every move is exactly reversible (byte-for-byte round-trip).
lawMoveReversible :: IsoMove -> OKLabI -> Bool
lawMoveReversible m x = apply (invert m) (apply m x) == x

-- | T3 — the identity move is a no-op on any colour.
lawIdentityIsNoOp :: OKLabI -> Bool
lawIdentityIsNoOp x = apply identityMove x == x

-- | T4 — 'sigmaMove' really is σ: it negates a and b and fixes L.
lawSigmaIsAMove :: OKLabI -> Bool
lawSigmaIsAMove (l, a, b) = apply sigmaMove (l, a, b) == (l, negate a, negate b)

-- | T5 — composition stays an exact isometry (the move group is closed under composition).
lawComposePreservesPairwiseDelta :: IsoMove -> IsoMove -> OKLabI -> OKLabI -> Bool
lawComposePreservesPairwiseDelta m2 m1 x y =
  distSqQ16 (apply (compose m2 m1) x) (apply (compose m2 m1) y) == distSqQ16 x y
